Allow the user to select the version of the docs

dimitri-tenstorrent opened the pull request:
### Ticket
#17433 

### Problem description

The users would like to see the documentation for several past versions but we currently only build the documentation for the latest commit on main. 

### What's changed

1. Added a UI selector to be able to switch between version of the documentation
The versions of the documentation are stored in https://github.com/tenstorrent/tt-metal/blob/dimitri/test-versioned-docs/docs/published_versions.json

The UI selector makes the HTTP request to get the list of the possible versions and adds the versions to the list for the user to select.

This required a change to the _layout.html for both `ttnn` and `tt-metalium`.

2. 
The deployment of Github pages (what we currently use to host documentation) needs to change to be done from the branch `gh-pages`. This is needed so that we store previous documentation and only add new versions to the branch as folders.

The proposed folder structure (e.g.):
```
v0.55.0/ttnn/index.html
v0.54.0/tt-metalium/index.html
```

We will need to change the settings for Github pages to be similar to the image below:

<img width="1201" alt="Screenshot 2025-01-31 at 1 24 50 PM" src="https://github.com/user-attachments/assets/c2693721-d8bb-4647-acba-35574d8aa341" />


3. We also need to adjust the package workflow to pass the version number to the docs building workflow and use a different Github actions to publish the folder.

### Additional Requirements
- [x] Display the version of the docs built next to the library name
- [x] Add a test to verify all links in the Installing.md and Readme.md
- [x] Add a test to make sure that after deploying docs the links to docs.tenstorrent.com remain accessible


### Checklist
- [x] Prefill the `gh-pages` branch with 3 versions of documentation: 55, 54, and 53
- [x] Switch the population of latest in the workflow
- [x] Flip the setting in the Github pages config (test with @tt-rkim )
- [x] Post commit CI passes 
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

dimitri-tenstorrent committed:
#17433: Build versioned docs in the gh-pages branch

dimitri-tenstorrent committed:
#17433: Fix the parameters for the action

dimitri-tenstorrent committed:
#17433: Simplify the workflow

dimitri-tenstorrent committed:
#17433: Implement versions in the sidebar for ttnn and tt-metalium

dimitri-tenstorrent committed:
#17433: Adjust how the versions are displayed

dimitri-tenstorrent committed:
#17433: Try another version for verification

dimitri-tenstorrent committed:
#17433: Pass the version from package and release workflow

dimitri-tenstorrent committed:
#17433: Only build docs on main

dimitri-tenstorrent committed:
#17433

dimitri-tenstorrent commented on docs/source/common/_templates/versions.html:
```diff
@@ -0,0 +1,45 @@
+</div>
+
+<script>
+const VERSIONS_URL = 'https://raw.githubusercontent.com/tenstorrent/tt-metal/refs/heads/dimitri/test-versioned-docs/docs/published_versions.json';
```
Don't forget to change this to `main` branch

ttmchiou replied:
this

dimitri-tenstorrent committed:
#17433: Fix linting

dimitri-tenstorrent committed:
#17433: Fix the error with version navigation

dimitri-tenstorrent committed:
Use the docs site url for testing

dimitri-tenstorrent committed:
#17433: Check the links for breakage

dimitri-tenstorrent committed:
#17433: Add runs on

dimitri-tenstorrent committed:
#17433: Add only Readme, Installing and actual documentation path

dimitri-tenstorrent committed:
#17433: Check all the links

dimitri-tenstorrent committed:
#17433: Move the link checker closer to the other doc check

dimitri-tenstorrent committed:
#17433: Fix the rebase conflict

dimitri-tenstorrent committed:
#17433: Update the paths of the links checkers

dimitri-tenstorrent commented:
See the version switching working on the staging server - https://tenstorrent-docs-test.github.io/tt-metal/v0.55.0/ttnn/index.html

ttmchiou reviewed (approved):
Remember to change to main branch

ttmchiou commented:
not sure if intentional but there's a UI glitch here @dimitri-tenstorrent ?
Might look nicer with a list too but thats just my opinion

![image](https://github.com/user-attachments/assets/9213bf90-a4f4-4a76-bb54-51181625c474)

ttmchiou commented:
Version 132.0.6834.160 (Official Build) (arm64)
on a macbook

dimitri-tenstorrent committed:
#17433: Fix styling bug that Michael reported

dimitri-tenstorrent commented on docs/source/common/_templates/versions.html:
```diff
@@ -0,0 +1,45 @@
+        data.versions.forEach(version => {
+            const dd = document.createElement('dd');
+            const link = document.createElement('a');
+            link.href = `https://tenstorrent-docs-test.github.io/tt-metal/${version}/${projectCode}/index.html`;
```
Don't forget to update the link to `tenstorrent.github.io` as well

dimitri-tenstorrent committed:
#17433: Add the version indicator to the project name

dimitri-tenstorrent committed:
#17433: Save the file changes

dimitri-tenstorrent commented:
> not sure if intentional but there's a UI glitch here @dimitri-tenstorrent ? Might look nicer with a list too but thats just my opinion
> 
> ![image](https://private-user-images.githubusercontent.com/156848643/411073561-9213bf90-a4f4-4a76-bb54-51181625c474.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMjYyMTgsIm5iZiI6MTczOTIyNTkxOCwicGF0aCI6Ii8xNTY4NDg2NDMvNDExMDczNTYxLTkyMTNiZjkwLWE0ZjQtNGE3Ni1iYjU0LTUxMTgxNjI1YzQ3NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMFQyMjE4MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ODdkYTVlMDhkZDBkMDRhYzIxNzczY2MyNGZlYjFiMzI4ZTA0YWU1MTBjNjZlMDQwZjU3MDBiOTE5YzQxZDdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.8TnH_v0e0fLByPS_2_wHgBeXFC2m0zMcO8yepNrRssE)

dimitri-tenstorrent commented:
Added the indicator for the version of documentation requested by @bkeith-TT 
![Screenshot 2025-02-10 at 5 49 34 PM](https://github.com/user-attachments/assets/2b3c3d9b-de14-4da7-99b9-a086691828c3)
Fixed the white bar issue reported by @ttmchiou 
![Screenshot 2025-02-10 at 5 49 46 PM](https://github.com/user-attachments/assets/afc21ffe-b40f-4dfd-94ff-bff7043ee9d9)

dimitri-tenstorrent committed:
#17433: Can't rename the check so just leave it as is

dimitri-tenstorrent committed:
#17433: Test the branch updates

dimitri-tenstorrent committed:
#17433: Prepare PR for merging

dimitri-tenstorrent committed:
Merge branch 'main' into dimitri/test-versioned-docs
